relay-server/src/endpoints/common.rs

```diff
@@ -307,12 +307,14 @@ fn queue_envelope(
 
         match state.envelope_buffer() {
             Some(buffer) => {
+                if !buffer.has_capacity().await {
```

Comment:
On second review, I think we should make this method sync, because we should avoid blocking the request handler on disk read / writes under all circumstances. One option to make it sync would be to change `GuardedEnvelopeBuffer::has_capacity` to a `try_lock`, and cache the resulting bool:

```rust
        match self.inner.try_lock() {
            Ok(guard) => {
                let c = guard.backend.has_capacity();
                self.cached_capacity.store(c);
                c
            },
            Err(_) => self.cached_capacity.load()
        }
```

Reply:
Yeah, this could be an idea, given the intrinsic delay of memory and disk usage calculation, adding one by caching it should not be a big deal. My main worry with this is that we can't guarantee updated values in any way if lock contention is arbitrarily high.

Reply:
Sounds acceptable to me. We can emit a counter metric to track how often we emit cached vs fresh capacity. Once we transform this to a service, the cached capacity will be updated differently, anyway.

Reply:
Yep, I added the metric because I am interested.